### PR TITLE
union to avoid void* casts

### PR DESCRIPTION
**Pull request by drystone**
@riataman I've only touched a few files but thought I'd explore some possibilities.

Some constness and a union so some parameters can be passed by value.

I'm not saying this is how it ought to be done ... I suppose I just don't like void *!

If there's anything here you think is worthwhile, I'm happy continue updating sources.

### CODE REVIEW

**Comment by riataman** [comment]
@drystone: The thing about passing a pointer to the value is that I could be modified. I guess if you want to be more safer it could be a pointer to its respective type. 

One of the things I want to do with this is be able to "redirect" all of return values, so I don't just need the value of the parameter, I need an actual pointer to it to be able to modify it!

**Comment by drystone** [comment]
Hmmm. Say we have `fn(int a, char *b)` `&a` is pointless as any modification is not going to get back to the caller. Similarly `&b` but modifications to `*b` are potentially useful and can be done directly without needing the address of `b` which will be on the call stack. Perhaps I'm misunderstanding something?

Also I think I'd be more inclined to separate the tracing and the modification into `trace_event(event_type, event_info)` and `struct event_info *modify_event(event_type, event_info)`

**Comment by drystone** [comment]
Ah - I see what you mean. You want to modify `a` and `b` before calling `real_fn(a, b)`. But I think it would be better style to implicitly use `real_fn(modified_event->param[0], modified_event->param[1])`

**Comment by drystone** [comment]
@riataman , sorry to go on. But what you're suggesting is that in `common.c` or somewhere else, in a GUI for example, you are going to have something like

```
if (!strcmp(func, "open") {
    /* do open modification */
} else if (!strcmp(fun, "connect") {
    /* do connect modification */
} ...
```
But this would be bad in terms of encapsulation and modularisation. I think modifications to `open` should be coded within `file.c` along with the `open` function. Unless you are thinking of using callbacks, which would be a bit pointless. Of course, there could be generalised modification funcs like `redirect_filename` but these should still be called from within our version of `open`.

**Comment by riataman** [comment]
I'm not going to do it like this:

if (!strcmp(func, "open") {
    /* do open modification */
} else if (!strcmp(fun, "connect") {
    /* do connect modification */
} ...

more like:

match_redirects (event_info)
{
redirect[] = get_redirects()

foreach(redirect)
{
    if (strcmp(redirect->function_name, event_info->function_name) == 0) {
       int parameter_num = redirect->parameter_num;
       void *redirect_value = redirect->redirect->value;

       event_info->param_values[parameter_num] = redirect_value;    
       break;
     }
}

and the config file would be something like:
# redirect,function_name,parameter_number,parameter_value
redirect,setuid,0,25

that would cause setuid's parameter 0 (uid) be set to 25. Same would apply to return values. Parameter could be modified in EVENT_TYPE_BEFORE_CALL and return values can be modified in EVENT_TYPE_AFTER_CALL.

I'm thinking that retrace_event(EVENT_TYPE_BEFORE_CALL) could also have a return value, to specify wherer or not the function be called at all. For example in the malloc redirection we have now we never actually call malloc() we just fake the return value. Same with getuid().

Basically all functions could be redirected without writing any special code for them. We can type check them using the type info we already have in the array. There's also going to be special cases, like connect where the redirection is going to be very complex, that should definitively stay in the connect's code itself. We could have a flag PARAMETER_FLAGS_DONT_REDIRECT to specify that it can't be redirected (most PARAMETER_TYPE_POINTER can't be redirected at all, other than to NULL I guess).

**Comment by drystone** [comment]
@riataman as you know I'm not entirely happy with the extent of this change. I think formalising the trace data/format is an excellent idea and this part of the patch is of real benefit. But you are also introducing a generic mechanism for modifying parameters before `real_func()` and/or the return value. This would be fine if we were using automatically generated shims for the overridden functions. In which case our `malloc.c` would simply contain hooks for pre and post modification functions. As it stands though, our `malloc.c` is a hand coded shim that should be as programmer friendly as possible.

I'd prefer to see separation of tracing and parameter modification and a more explicit code style for parameter modification. It's not nice to have variables updated as a side effect of calling a function and such side effects are a common cause of problems.

For an (unrealistic) example, in your scheme:
```
void *malloc(size_t size)
{
    int r;
    struct rtr_event_info ei;
    ei.params = {&size};  /* all cast to void ** */
    retrace_event(&ei); /* may modify size as a side effect */
                                   /* by the malloc-agnostic modification function */
                                   /* and who does freeing of any allocated buffers for modified data? */
    size = my_malloc_mod(); /* loses modification but an easy mistake to make */
    r = real_malloc(size);
    retrace_event(&ei); /* same again - r may be modified as a side effect */
    r = malloc_fuzz();  /* and modification lost ... */
    return r;
}
```
I think it would be better to have the more explicit
```
void *malloc(size_t size)
{
    struct rtr_call_info ci, mci;
    ci.params[0].intval = size;  /* typed */
    rtr_log_call(BEFORE, &ci);
    if (rtr_redirect_args(&ci, &mci) /* no voodoo */
        size = mci.params[0].intval;
    else
        size = my_malloc_size;
    ci.retval.pvoid = real_malloc(size);
    rtr_log_call(AFTER, &ci);
    return ci.retval.pviod;
}
```
Even this has problems as the before and after logging may get separated in the output in multithreaded programs.

**Comment by drystone** [comment]
@riataman sorry, I meant for that to go into riboseinc/retrace
